models/classes/event/QtiTestDeletedEvent.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2023 (original work) Open Assessment Technologies SA.
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoQtiTest\models\event;
+
+use oat\oatbox\event\Event;
+
+class QtiTestDeletedEvent implements Event
```

Comment:
I am not sure if we needed this event as we have `oat\taoTests\models\event\TestRemovedEvent` already.

Reply:
This event carries the URIs for deleted tests, deleted items and related assets. `TestRemovedEvent` only carries the URI for the deleted test.

If we use `TestRemovedEvent`, we would:
- Trigger an event per deleted test (usually a single one, but maybe more if there are more tests with the label used in the import). 
- Use the `ElementReferencesExtractor` (from QtiTest) in MediaManager to get the asset refs (current changes already add a dependency in MM pointing to QtiTest). We would need to do that before actually removing the test and the items (I think once the test is deleted the refs are gone).

The event is currently triggered by taoTest (`deleteTest()` method).

Therefore, we'd need to trigger `TestRemovedEvent` before the test is actually removed in order to get the list of assets from MM, or extend `TestRemovedEvent` to carry the new info and pass that info to taoTest method `deleteTest()` to include it in the event.

I find it easier to add this new event to hold the referenced resources, since we already had logic here to determine the referenced items (items themselves are already deleted by QtiTestService).

Reply:
have you considered a child class from TestRemovedEvent @hectoras ?

Reply:
We may do that, but this method carries an array of test URIs, so we'd need to trigger an event per test instead of a single one.

Reply:
We have to keep in mind the confusion that might cause/bring on having multiple similar events for the same thing just because they are called in a different place, etc. 

If it is not a big deal, I would suggest to: 

- add missing parameters for `TestRemovedEvent` (maybe a param called `array $extendedPayload`, so this is extensible). 
- Trigger this event before deleting the test / item classes.

I am not sure adding new `duplicated` event is a solution when we already have one.

If extending the event is too complicated, then create a new event focused on mediaRemoval instead. 

```
class CleanUpMediaEvent
```

That is focused on deletion...

But I would not have 2 events for Test Deletion, cause it will be confused for developers.